lua/nvim-lsp-installer/ui/display.lua

```diff
@@ -175,18 +185,12 @@ function M.new_view_only_win(name)
     end
 
     local draw = process.debounced(function(view)
-        local win_valid = vim.api.nvim_win_is_valid(win_id)
-        local buf_valid = vim.api.nvim_buf_is_valid(bufnr)
+        local win_valid = win_id ~= nil and vim.api.nvim_win_is_valid(win_id)
+        local buf_valid = bufnr ~= nil and vim.api.nvim_buf_is_valid(bufnr)
         if not win_valid or not buf_valid then
             -- the window has been closed or the buffer is somehow no longer valid
             unsubscribe(true)
-            if win_valid then
-                vim.api.nvim_win_close(win_id, true)
-            end
-            if buf_valid then
-                vim.api.nvim_buf_delete(bufnr, { force = true })
-            end
-            redraw_by_winnr[win_id] = nil
+            M.delete_win_buf(win_id, bufnr)
```

Comment:
I'm not sure why this call is here. Can we even delete a window with an invalid id?
However, I think there's more to it than disabling this function alone.

Reply:
I'd say there's no specific reason other than making sure to do an exhaustive clean up should it ever try to draw/render when the buffer-window pair is invalid to avoid potential lingering windows/buffers.

Reply:
I think I found the bug, please try the latest commit and let me know if it works for you.

Reply:
Hmm it kinda defeats the purpose I think. We need to make sure to delete the buffer (if it's still valid), otherwise you can split it into a new window. The window management is pretty solid I think, thanks to the autocmds and whatnot. This is mostly about ensuring that the buffer follows suite (deletes itself when the popup window closes, or a different way of expressing it - doesn't allow itself to be loaded into another window than the floating window).

What about removing this delete call from the render function (seeing as there is no strong reason to have it in here in the first place), and keeping it as it was (delete both win and buf)? Would that work?

Reply:
Okay, at least I feel like there's more info to go about debugging this now. Here's what happens when you try to list the buffers (notice how a new window **should** wipeout everything). You can also add `previewer=false` to remove the preview warning.

![TvazZ90yRF](https://user-images.githubusercontent.com/59826753/135115688-fb4e14d9-6608-46b9-8b1d-e2690f5233af.gif)

Reply:
Hmm that's probably just telescope gathering all buffers before opening a new window. When the telescope window is later opened our buffer will have been wiped. If that's the case I feel like it's not related to this plugin at all?

Try doing a vertical split `:vsp` when the floating window is open. Since the buffer will be loaded into a new window before the floating window is closed, it won't delete itself (`bufhidden` behavior is never triggered). The next time it tries to render it will suspend because the original floating window handle is no longer valid. This is why I want the `delete_win_buf` function to delete both the provided window and buffer, because it will be called on `WinLeave`, which will happen in due time. Due to this call to `delete_win_buf` the buffer will actually be deleted, and the newly opened window will be closed.

Reply:
I got stuck trying my hand at this (https://github.com/kylo252/nvim-lsp-installer/tree/display-refactor), but it's way too complicated. I should've started trying to understand how it's used before diving head first into such adventures 😅 

I can't tell you if the current architecture is good or not, mostly because I don't fully get it. Maybe my problem is that I'm treating it as a singleton or that it should be a singleton. I can't help but feel that you have some redundant state management, when you can just instantiate a new object. 

Instead there's a display object that needs to go through this confusing lifecycle: 

🟢 `new()` 
➡️ `view()`
➡️ `init()` "wait, what does `new()` do then? "
➡️ `open()` "open what? what did `view()` do before? how come you can call `view()` before `init()`?"

Some of this might be easily remedied with some renaming, well, I hope so!

ref: https://github.com/williamboman/nvim-lsp-installer/blob/89b5a04fb5bce23cdd77e5b9f2e175f60404e429/lua/nvim-lsp-installer/ui/status-win/init.lua#L236-L268

Reply:
Haha yes it's all a bit thrown together at the moment.

1. Make it work <--- still here :D
1. Make it right
1. Make it fast 

The UI framework I built here basically have 3 main constructs:

1) The stateless "view" function
The view function is responsible for returning an in-memory representation of what should be rendered. This representation is basically just a tree structure of nodes (I believe currently, leaf nodes are either always of the `Ui.Node` or `Ui.HlTextNode` types). It holds no state by itself, but instead receives the state as an argument. This is what's registered when calling `.view(fn)`.

2) The state container
The state container is a simple container for a piece of mutable state. A state container allows for a) mutating the current state, b) retrieving the current state, and c) subscribing to state changes (this is what schedules the next render). The state container is set up when calling `.init(initial_state)`.

3) The "display" object
The display is basically a representation of a buffer-window pair. The window can be any valid neovim window, and I can imagine there being multiple different variants of it (floating, "normal", preview, etc.). What type it is should have no bearing on what can be rendered - but the renderer might do things differently depending on what type it is. The display object is responsible for translating the in-memory representation provided by the view function into `vim.api.nvim_buf` calls (to set lines, highlights, extmarks, etc.). It's basically responsbile for making sure 1) and 2) has a graphical outlet.

Reply:
> I got stuck trying my hand at this (https://github.com/kylo252/nvim-lsp-installer/tree/display-refactor), but it's way too complicated. I should've started trying to understand how it's used before diving head first into such adventures 😅

Haha I think it looks great though :+1:! I'm actually pretty satisfied with how this PR looks like right now - are you still getting the segfaults or should we just go ahead and merge? I've tested on:

- macOS (recent nightly)
- Windows (recent nightly)
- Ubuntu x64 (recent nightly)
- Alpine Linux arm (recent nightly)

And it all works fine!

I'd be more than happy to receive another PR to improve this piece of code. What do you say?

Reply:
> The UI framework I built here basically have 3 main constructs

This will take some time to process, I'm gonna to have to dig into it at some point if you decide to go with the refactor (in a separate PR ofc, it's too much work to jumble it with this one 😄)

> are you still getting the segfaults or should we just go ahead and merge? I've tested on:

Yeah I'm still getting the segfault unfortunately, by basically only pressing `<c-w>` and I've tried it on multiple systems with the same results. I'm confused how you're not encountering this.

Reply:
Do you have any other plugins enabled? Or can you repro on clean install with only this plugin loaded?

Reply:
Damn, you're right! I just tried it with a `minimal_init.lua` and it seems to be okay. 🎉

So something is up with a configuration or a plugin that I have. That should hopefully make debugging much easier.

Reply:
My bet is on https://github.com/folke/which-key.nvim :D

Reply:
It was actually which-key 😢 

I'm trying to replicate this with other plugins. It seems that Telescope knows how to deal with this, whereas Packer sends a wild error message and destroys both windows.

I have a feeling this could be avoided by checking the filetype, but that only works if the info panel was trying to hijack which-key, and not the other way around.

Reply:
Does it still crash with the latest commit ca446fe59b012b8e18e2b6d1aa74a94e37f3585b?

Reply:
1486b3700c25f771c4bfb01a3400d94cc5a23b3c seems to fix it! Weird one, not gonna bother investigating further haha

Reply:
I was literally going to add a very similar commit, `vim.schedule` is the way!